### ISSUE TITLE
OC v059 Misidentifies Cloned Drive after Applying SecUpd2020-003

### ISSUE THREAD

**Issue by dakanji** [acidanthera/bugtracker#957]
While I have seven HDs in my cMP 3,1, OC is setup to give me two APFS options to boot from:
1. "Macintosh HD": My main working drive with Mojave installed
2. "Clone": A daily backup of Macintosh HD using Carbon Copy Cloner

Ever since installing the recently released SecUpd2020-003Mojave, OC has identifying "Clone" as "Macintosh HD". I.E., I get two options to both from ... "Macintosh HD" and "Macintosh HD". 

See log extract below:
![BootLogCropped](https://user-images.githubusercontent.com/1128876/83937404-ec71ae80-a7d4-11ea-8b25-c56176b3dae2.png)

Once I get into the OS, "Clone" is labelled correctly. I use Refind to chainload OC and in it is also labelled correctly in Refind. Issue only appears to exist in OC.
 
Full Log File: [opencore-2020-06-06-052943.txt](https://github.com/acidanthera/bugtracker/files/4739425/opencore-2020-06-06-052943.txt)

I suppose I could write a script to rename the files files it is getting the label from as a post flight script in CCC (has to be on each clone as if done once, they would be overwritten on the next backup) but that would be treating the symptom especially since macOS and Refind are able to properly get the label ... presumably from somewhere else.

PS. As title says, this only started after applying SecUpd2020-003Mojave. Could be related to [THIS ISSUE](https://bombich.com/blog/2020/05/27/bug-in-macos-10.15.5-impacts-bootable-backups-weve-got-you-covered)

CCC seems to have changed something in their update but this seems to only impact OC.

**Comment by vit9696** [comment]
That's a macOS bug unfortunately. It started to appear with APFS. See https://github.com/acidanthera/bugtracker/issues/886. What you need to do is report to Apple.

**Comment by dakanji** [comment]
> What you need to do is report to Apple

Fair enough and understand the "WontFix" decision but worth throwing out there that given that Refind is not affected by this, and that not withstanding the bug from Apple, perhaps OC could be a bit more resilient in this aspect.

**Comment by dakanji** [comment]
**Update** 
This line: `03:717 00:014 OCB: Trying to get label from \3231689A-AEC2-4860-83A7-305B05C2DB4E\System\Library\CoreServices\.disk_label.contentDetails` should have returned the correct name as it is present it that file.

The only item of note is that the file tried here: `03:717 00:014 OCB: Trying to get label from \3231689A-AEC2-4860-83A7-305B05C2DB4E\System\Library\CoreServices\.contentDetails` does not exist.

So basically, the second of the files checked has the correct label in it. 

Worth looking into further?

**Comment by vit9696** [comment]
> Fair enough and understand the "WontFix" decision but worth throwing out there that given that Refind is not affected by this, and that not withstanding the bug from Apple, perhaps OC could be a bit more resilient in this aspect.

rEFInd has a lot of other issues determining boot entry names, so it is a trade-off of using one method or another, which additionally concerns the performance.

> This line: 03:717 00:014 OCB: Trying to get label from \3231689A-AEC2-4860-83A7-305B05C2DB4E\System\Library\CoreServices\.disk_label.contentDetails should have returned the correct name as it is present and correct in that file.

That is exactly the issue I have just explained to you, but you misunderstood it:

- The fact you see two prints means that the first file one was not found, and the second one was used.
- If both files had not been present, then the volume name would have been used (similarly to rEFInd), and no issue would have happened.
- However, the issue here is that the file was places by Apple, forcing us to use it, but it was **not** updated or rather was **improperly** updated by Apple during the update installation.
- Later it was then fixed afterwards, again by Apple, which led to the correct name, but initially it indeed contained `Macintosh HD` just like now it has Clone.

If you entered native Apple Boot Picker during this procedure you would have likely faced the same issue. Since you normally you do not enable the picker at all (and OpenCore is not an exception with ShowPicker = NO being recommended), this issue is probably largely unnoticed and low-pri in Apple.

There is one corner-case here, which makes things even worse. While OpenCanopy (if you enable that in PickerAttributes) and Apple Boot Picker use graphical labels (`.disk_label_2x` / `.disk_label` files), textual picker uses text labels (`.disk_label.contentDetails` / `.contentDetails`) instead. As a result there is room for technically having different data in them. 

Guess what? Apple sometimes does leave them in desync, i.e. a graphical entry has one string, and a textual label has another string. For instance, this is what happened to HFS+ recoveries some time ago, and as a result we even built special workarounds here in [BootEntryInfo.c#L557-L568](https://github.com/acidanthera/OpenCorePkg/blob/afd9823/Library/OcBootManagementLib/BootEntryInfo.c#L557-L568).

Unfortunately for your case we cannot really add another similar hack for two reasons:

- the volume in charge may be encrypted and not available (i.e. we only see the `Preboot` and cannot get the volume label in the first place), which means there is no way to build a reliable solution for all cases
- the issue happens rarely enough not to care and providing a partial workaround will add unnecessary slowdowns.

Took some time to create this write-up, hope it is appreciated.

**Comment by dakanji** [comment]
> Took some time to create this write-up, hope it is appreciated

Most definitely

> The fact you see two prints means that the first file one was not found, and the second one was used

Yes, I found the logic in `Library/OcBootManagementLib/BootEntryInfo.c` and appreciate the issue better following your explanation.

I see two possible ways forward for me and others in my situation:

1. Delete `/System/Library/CoreServices/.disk_label.contentDetails` and let OC default to the Volume name as you mentioned (probably not a good idea)
2. Create `/System/Library/CoreServices/.contentDetails` and hopefully OC picks this up and uses it.

Anyway, thanks again. Will leave it here and find a workaround.

**Comment by vit9696** [comment]
Yeah, but should really file a radar in any case via http://feedbackassistant.apple.com

**Comment by dakanji** [comment]
> ... should really file a radar in any case

Sure. Just need to find an Apple specific way of reproducing the issue.

> Later it was then fixed afterwards, again by Apple, which led to the correct name, but initially it indeed contained Macintosh HD just like now it has Clone.

Must confess I didn't (and still don't) quite understand this as I thought "if it now has the correct entry, why is this not being used when it is read?".

Anyway, I built OC with an additional debug flag `DA_DBG` to better see what is going on there: 
![Debug](https://user-images.githubusercontent.com/1128876/83940495-36668e80-a7ed-11ea-8c0a-30c9543a3252.png)

None the wiser but can see something is going on with the `McnohH` value which crops up in the first one and apparently translates to "Macintosh HD". I presume in my limited understanding that this is coming from somewhere else. 

Moving on, will test now the workarounds and figure out how to get a flag to Apple that will not be dismissed out of hand considering that I am on an unsupported machine.

Thanks again for the insights.

**Comment by vit9696** [comment]
You used `%s` to print ASCII strings. They need to be printed with `%a` in UEFI world (do not ask me why).

**Comment by dakanji** [comment]
For the benefit of anyone stumbling upon this from web search etc, putting this in before the notoriously grumpy Acidanthera guys slam this shut :grin:

The first of the two possible ways forward outlined in my earlier comment actually works well and reliably. The only issue is that the Disk(s) will show up as `Preboot` in the OC BootPicker. The second option DOES NOT WORK. Still always get `Macintosh HD`. 

I could have stuck with Option 1 since it gives me `Preboot` & `Macintosh HD` which is OK to go with since I only have two and could therefore ID them easily but I decided instead to patch the `InternalGetAppleDiskLabel ` function in `Library/OcBootManagementLib/BootEntryInfo.c` and everything is no working as before as of June 2020. 

Code below:
```
InternalGetAppleDiskLabel (
  IN  EFI_SIMPLE_FILE_SYSTEM_PROTOCOL  *FileSystem,
  IN  CONST CHAR16                     *BootDirectoryName,
  IN  CONST CHAR16                     *LabelFilename
  )
{
  CHAR16   *DiskLabelPath;
  UINTN    DiskLabelPathSize;
  CHAR8    *AsciiDiskLabel;
  CHAR16   *UnicodeDiskLabel;
  UINT32   DiskLabelLength;

  DiskLabelPathSize = StrSize (BootDirectoryName) + StrSize (LabelFilename) - sizeof (CHAR16);
  DiskLabelPath     = AllocatePool (DiskLabelPathSize);

  if (DiskLabelPath == NULL) {
    return NULL;
  }

  UnicodeSPrint (DiskLabelPath, DiskLabelPathSize, L"%s%s", BootDirectoryName, LabelFilename);
  DEBUG ((DEBUG_INFO, "OCB: Trying to get label from %s\n", DiskLabelPath));

  UnicodeDiskLabel = NULL;
  if (OcStriStr (DiskLabelPath, L"3231689A-AEC2-4860-83A7-305B05C2DB4E") != NULL) {
      AsciiDiskLabel = NULL;
      UnicodeDiskLabel = GetVolumeLabel (FileSystem);

      if (OcStriStr (UnicodeDiskLabel, L"PreBoot") != NULL) {
          UnicodeDiskLabel = L"Clone";
      }

  } else {
      AsciiDiskLabel = (CHAR8 *) ReadFile (FileSystem, DiskLabelPath, &DiskLabelLength, OC_MAX_VOLUME_LABEL_SIZE);
  }
  FreePool (DiskLabelPath);

  if (AsciiDiskLabel != NULL) {
    UnicodeDiskLabel = AsciiStrCopyToUnicode (AsciiDiskLabel, DiskLabelLength);
    FreePool (AsciiDiskLabel);
  }

  if (UnicodeDiskLabel != NULL) {
    UnicodeFilterString (UnicodeDiskLabel, TRUE);
  }

  DEBUG ((DEBUG_INFO, "OCB: Disk Label: %s\n", UnicodeDiskLabel));

  return UnicodeDiskLabel;
}
```
 
Note that this is specific for my disk and you will need to change `3231689A-AEC2-4860-83A7-305B05C2DB4E` and `Clone` to match the disk you are having issues with.

It can be if .... else if .... else as needed.
